### ISSUE TITLE
Proposal: Ensure.That() (name tbd) which only raises AssertionError, doesn't fail test

### ISSUE THREAD

**Issue by vivainio** [nunit/nunit#3367]
This is proposal to address problem described at https://github.com/nunit/nunit/issues/3242

We have lots of code that relies on nunit 3.5 Assert behavior, i.e. it raised exception but didn't fail the test. We caught the exception and retried in some cases.

We are blocked on upgrading NUnit because of this. Our temporary solution is "forking" the Assert class so that it only raises an exception like it used to in NUnit 3.5.

NUnit could facilitate this migration by creating a new, supported Ensure static class which does what the old Assert did. We would prefer if this was maintained in upstream NUnit instead of maintaining our own forked class.

**Comment by CharliePoole** [comment]
It seems odd to provide the same feature (asserting) under two names, with different implementations. And, if we did that, we would have to promise not to change the implementation. It seems more logical to me to provide a new feature: try an assertion and return true or false depending on the result.

What if we had something like `Assert.Try` or `Check.That` returning a boolean?

**Comment by vivainio** [comment]
Boolean wouldn't work because you still want the "rich" failure messages following the NUnit constraints engine. And optimally, there would not be need to "check return value and raise exception", because then you end up creating a wrapper class that you use everywhere, and would still need to maintain when NUnit is updated. 

Check.That, Check.AreEqual are ok as names of course. But it would be best if projects still stuck on NUnit 3.5 could just do a global search-and-replace (both Assert. and Does. would be converted to Check.). Think of it as a compatibility feature.

**Comment by CharliePoole** [comment]
Hmmm... What if it returned and AssertionResult? This is currently an internal Type, of course, but it might reasonably be exposed.

**Comment by vivainio** [comment]
Closing this as a "solved problem from our side", event if NUnit changes could still be done in the future.

In the end, we decided to decouple our "page object" (i.e. non-test-case) assertions from NUnit, i.e. we switched them over to NFluent (and as such they don't take a reference to NUnit anymore, and won't trigger forced test failures).

Going forward, it could be useful to ship assertions and constraints in a separate nuget package.
